File: C Sharp/Blink/Blink/Core/SendPacket.cs
using Net.Qiujuer.Blink.Listener;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace Net.Qiujuer.Blink.Core
{
    public abstract class SendPacket : BlinkPacket
    {
        public SendListener Listener { get; private set; }
        private bool mCanceled;
        private BlinkConn mBlinkConn;

        public SendPacket(byte type, SendListener listener)
            : base(type)
        {
            Listener = listener;
        }

        /// <summary>
        /// Cancel the packet to send
        /// If the packet on sending you can't cancel it
        /// But you can cancel sending notify callback
        /// </summary>
        public void Cancel()
        {
            mCanceled = true;
            if (mBlinkConn != null)
            {
                mBlinkConn.Cancel(this);
                mBlinkConn = null;
            }
        }

        /// <summary>
        /// Get the packet iscanceled
        /// </summary>
        /// <returns>Is Canceled</returns>
        public bool IsCanceled()
        {
            return mCanceled;
        }

        /// <summary>
        /// Set The BlinkConn to Cancel from queue
        /// </summary>
        /// <param name="blinkConn">BlinkConn</param>
        /// <returns>SendPacket</returns>
        public SendPacket SetBlinkConn(BlinkConn blinkConn)
        {
            mBlinkConn = blinkConn;
            return this;
        }

        /// <summary>
        /// On Sender send the packet call this to send packet info
        /// The bytes in 0~32767 size
        /// </summary>
        /// <param name="buffer">Send buffer</param>
        /// <param name="index">Buffer start index</param>
        /// <returns>Read to buffer count</returns>
        internal virtual short ReadInfo(byte[] buffer, int index)
        {
            return 0;
        }

        /// <summary>
        /// Sender read some data to send
        /// </summary>
        /// <param name="buffer">Buffer</param>
        /// <param name="offset">Buffer offset</param>
        /// <param name="count">Buffer count</param>
        /// <returns>Read to buffer count</returns>
        internal abstract int Read(byte[] buffer, int offset, int count);

    }
}

File: Java/Blink/library/src/main/java/net/qiujuer/blink/core/SendPacket.java
/*
 * Copyright (C) 2014 Qiujuer <qiujuer@live.cn>
 * WebSite http://www.qiujuer.net
 * Created 04/16/2015
 * Changed 04/19/2015
 * Version 1.0.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.qiujuer.blink.core;

import net.qiujuer.blink.listener.SendListener;

/**
 * Blink SendPacket
 */
public abstract class SendPacket extends BlinkPacket {
    private final SendListener mListener;
    private boolean mCanceled;
    private BlinkConn mBlinkConn;


    public SendPacket(byte type, SendListener listener) {
        super(type);
        mListener = listener;
    }

    public SendListener getListener() {
        return mListener;
    }

    /**
     * Cancel the packet to send
     * If the packet on sending you can't cancel it
     * But you can cancel sending notify callback
     */
    public void cancel() {
        mCanceled = true;
        if (mBlinkConn != null) {
            mBlinkConn.cancel(this);
            mBlinkConn = null;
        }
    }

    /**
     * Get the packet iscanceled
     *
     * @return Is Canceled
     */
    public boolean isCanceled() {
        return mCanceled;
    }

    /**
     * Set The BlinkConn to Cancel from queue
     *
     * @param blinkConn BlinkConn
     * @return SendPacket
     */
    SendPacket setBlinkConn(BlinkConn blinkConn) {
        mBlinkConn = blinkConn;
        return this;
    }

    /**
     * On Sender send the packet call this to send packet info
     * The bytes in 0~32767 size
     *
     * @param buffer Send buffer
     * @param index  Buffer start index
     * @return Read to buffer count
     */
    public short readInfo(byte[] buffer, int index) {
        return 0;
    }

    /**
     * Sender read same data to send
     *
     * @param buffer Buffer
     * @param offset Buffer offset
     * @param count  Buffer count
     * @return Read to buffer count
     */
    public abstract int read(byte[] buffer, int offset, int count);
}
